Allow construction of async stream objects without triggering async operations

vjpai opened the pull request:
This is important because it enables an object to be fully constructed and available before another thread sees its CQ event trigger. Otherwise there needs to be explicit application-level synchronization in place to avoid that problem.

Although technically this PR isn't required for async unary calls (since those don't have a CQ tag at construction), I've applied it there as well for consistency.

Fixes #11906

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                            cpu_time    real_time
-----------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/32k/1                      +4%         +5%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/32k/2                      +4%         +4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/256k                   +13%        +12%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32k                 +4%         +4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/256k/1/0  +11%        +11%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/32k/2/0   +4%         +4%
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                     cpu_time    real_time
--------------------------------------------------------------------------------------------  ----------  -----------
BM_IsolatedFilter<DummyFilter, NoOp>                                                          +5%         +5%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32k                          +16%        +15%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/32k/2/0      +7%         +8%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/1/0         +7%         +7%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/0         +4%         +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/1         +4%         +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/256k/1/0  +6%         +6%
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                              cpu_time    real_time
-------------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/256k/2                       -9%         -9%
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/256k/1                    -9%         -9%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/32k/1/0     -4%         -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/0  -8%         -8%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/1  -5%         -5%
```

vjpai commented:
@a11r : how do the new API names look to you?

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                               cpu_time    real_time
--------------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/256k/1                        +8%         +7%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/32k/2                         +4%         +4%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/256k/1               +5%         +5%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/32k                       +13%        +13%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32k                    +5%         +5%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/256k/1/1     +7%         +7%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/256k/1/1  +4%         +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/256k/2/1  +6%         +6%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/0   +12%        +12%
```

vjpai commented:
This PR is still broken. I need to reach into C core and change the behavior of `grpc_alarm_create` (or create an alternative `grpc_alarm_create_unraced` that takes a `grpc_alarm**` as its argument and fills it in)

vjpai commented:
I will separate out the alarm component of this from the C++ streaming objects part since the alarm portion requires core work as well.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

vjpai commented:
Cc @slash-lib

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

yang-g commented on include/grpc++/impl/codegen/async_stream.h:
```diff
@@ -165,18 +168,20 @@ class ClientAsyncReader final : public ClientAsyncReaderInterface<R> {
                                    CompletionQueue* cq, const RpcMethod& method,
                                    ClientContext* context, const W& request,
-                                   void* tag) {
+                                   bool start, void* tag) {
```
Please update the comment above to include `start`.
We may want to add an assert to say when you set `start` to true you have to pass in a `nullptr` as tag to prevent mis-use?

vjpai replied:
I added the comments and asserts requested, plus also asserts on each of the Read/Write/Finish functions to make sure that started_ has already happened by that time (either through the constructor or StartCall)

yang-g commented on include/grpc++/impl/codegen/async_stream.h:
```diff
@@ -35,6 +35,9 @@ class ClientAsyncStreamingInterface {
   virtual ~ClientAsyncStreamingInterface() {}
 
+  /// Start the call that was set up by the constructor
+  virtual void StartCall(void* tag) = 0;
```
We will need a big warning comment here saying this is only needed when a prepare call was used to create the reader object. I think it might be helpful to buffer the start bit in the object and assert fail if it is true.

vjpai replied:
Excellent suggestion, I did this.

yang-g commented:
LGTM

yang-g commented:
If we want to advise users to use this new API, we should also update the async examples and tests and make it clear in the documents that the new API should be used.

vjpai commented:
Yes, I will plan to document this. The reality is that users can continue to use the existing API, particularly for single-threaded codes or if they like dealing with workarounds. But we can update the examples to make it clear that this should be the new normal. Now or in a follow-on PR?

vjpai commented:
Ok, I have updated the examples. As for the tests, there are three categories:

1. async/generic/hybrid_end2end_test : not directly affected by API workarounds since single-threaded
2. thread_stress_test : not directly affected by API workarounds since unary only (where the new API only exists for consistency, not functionality)
3. qps tests: changed in this PR. This PR allows the reduction of a lock at object construction.

vjpai commented:
Rebased and squashed; diff is unchanged.

vjpai commented:
Jenkins, test this please

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

vjpai commented:
Jenkins, test this please

vjpai commented:
Jenkins, test this please

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.new: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                              atm_add_per_iteration    nows_per_iteration
---------------------------------------------------------------------  -----------------------  --------------------
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/16M/2                           -4%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/2M/2               -5%                      -5%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/2M               -5%                      -6%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/2M/2M                  -5%                      -6%
```

vjpai commented:
Test failures in CQ (to be resolved by #12564)
Test failures #12307 #12138 #12407 #12547
